### ISSUE TITLE
Restart admission credits ALL unattributed node memory to the restarting deployment, so a fine-tune job makes the node look empty

### ISSUE THREAD

**Issue by kreuzhofer** [kreuzhofer/dgx-manager#118]
Found while running an unowned probe container on spark-03 for #116, but **it does not need one** — a running fine-tune job reaches the same state through ordinary product use.

## The mechanism

`checkVllmVramAdmission` derives the node's used memory from the live metric, which is correct and not the problem:

```ts
// admission/vram.ts:168
const vramUsedMB = latestMetric?.vramUsed || 0;
```

On a restart it then computes how much of that the restarting deployment is about to free:

```ts
// admission/vram.ts:170-176
if (excludeDeploymentId) {
  const otherFootprintMB = conflicts.reduce(
    (sum, c) => sum + (c.vramActualMB ?? c.vramEstimateMB ?? 0), 0,
  );
  reclaimableMB = Math.max(0, vramUsedMB - otherFootprintMB);
}
```

and `computeVramShortfall` subtracts it: `const vramUsed = Math.max(0, snapshot.vramUsedMB - reclaimable)` (`vram.ts:96`).

So **`reclaimable` is defined as "everything not attributable to another deployment row"**. Any memory on the node that is not a `Deployment`/`ClusterNode` row gets treated as the restarting deployment's own, about to be freed. With no other rows present, `reclaimable == vramUsedMB`, `vramUsed` computes to **0**, and the node reads as completely empty no matter how full it is. The restart is admitted, launches, and contends. On GB10 unified memory that is the OOM-kill path from #92.

**Fresh deploys are unaffected** — `reclaimableMB` stays 0 when `excludeDeploymentId` is unset, so the live metric is used as-is and an over-commit is correctly refused.

## Three ways to get unattributed memory, without doing anything unusual

1. **A running fine-tune job.** `loadConflicts` queries only `prisma.deployment` and `prisma.clusterNode`; `admission/vram.ts` contains no reference to fine-tune at all. `FineTuneJob` has a `nodeId` and a `status` but no VRAM columns, so a training container holding tens of GiB is **invisible to admission**. Restarting any deployment on a node that is training will read that node as emptier than it is. This is the path that needs no mistake by anyone.
2. **An orphaned container after node offboarding.** `routes/nodes.ts:55` does `prisma.deployment.deleteMany({ where: { nodeId } })`. Rows go; a surviving container's memory does not.
3. **Any manually-run container** — a probe, a hand-started benchmark, someone else's work. This is how I hit it.

## Second, smaller defect: the refusal cannot explain itself

`loadConflicts` also builds the **conflict list shown to the user**. When unattributed memory causes a *fresh* deploy to be refused, the 409 arrives with an empty conflict list: something is holding the memory and the product cannot say what. That is a concrete instance of #101, and it is what turns this into an archaeology session rather than a message.

## Why it is written this way

Deliberately, per the comment at `vram.ts:159-167` and issue #1: deriving reclaimable from *node metric minus other rows* avoids depending on whether a deployment's own `vramActual` is stored per-node or summed across a cluster. That reasoning is sound. The unconsidered case is that the subtraction silently assumes **every** byte on the node belongs either to another deployment or to the restarting one — and the node is not that tidy.

## Suggested direction

Bound `reclaimable` by what the restarting deployment is actually known to hold, rather than by everything unexplained:

```ts
reclaimableMB = Math.min(
  self.vramActualMB ?? self.vramEstimateMB ?? 0,
  Math.max(0, vramUsedMB - otherFootprintMB),
);
```

That keeps #1's intent (a restart is not rejected for memory it is about to free) while making unattributed memory count *against* the deploy instead of for it — failing closed. The cluster-attribution problem #1 worried about becomes a reason to prefer the `max` of per-node candidates, not a reason to treat the whole node as reclaimable.

Whatever the fix, `computeVramShortfall` is already pure and property-testable, so the invariant is stateable directly: **for any snapshot, `vramUsed` after reclaim must never be lower than `vramUsedMB` minus the restarting deployment's own recorded footprint.**

## Also worth deciding separately

Whether fine-tune jobs should appear in the conflict list at all. They occupy nodes, they are visible to the manager, and today no admission path accounts for them.

**Comment by kreuzhofer** [comment]
> *This was generated by AI during triage.*

## Agent Brief

**Category:** bug
**Summary:** Bound a restart's VRAM reclaim by the restarting deployment's own authorised share, so unattributed node memory counts *against* the restart instead of for it — and let the refusal name a fine-tune job as a holder.

### Decisions taken during triage

These were settled in a grilling session and are **not open for re-litigation**. An implementation that reopens one of them is wrong even if it passes the tests.

1. **The manager measures nodes, never deployments.** A per-deployment memory figure is always a *claim*, never a measurement. Per-process/per-container GPU attribution at the agent was considered and rejected — it is mechanically possible (`nvidia-smi --query-compute-apps` already supplies the node total on GB10) but can never be complete, because on a unified-memory pool the memory that actually kills a deploy is often held by a process with no CUDA context at all (the 126 GiB download in #92). A design that depends on attribution being complete has this same bug with a smaller trigger.
2. **The bound is the authorised share, derived at admission time**, not read from the stored `vramEstimate` column. `round(node.vramTotal × gpuMemUtil)` is well-defined per node, needs no migration or backfill, and dissolves the cluster-attribution worry that led #1 to the node-metric trick. (The stored column is `null` for every multi-node deploy — the estimate is only computed on the non-cluster branch — so bounding by it would 409 every GLM-class cluster restart.)
3. **The reclaim uses the deployment's *saved* gpu_memory_utilization; the request uses the caller's.** They are two different numbers. Restarting at a higher share must not widen what the deployment is credited for.
4. **The `otherFootprintMB` subtraction is deleted, not kept as a second term.** With the reclaim capped by the authorised share, it adds no safety and imports a defect: the per-deployment figure it sums is the *whole node's* reading for vLLM/dgxrun runtimes, so two deployments on one node over-count each other, clamp the reclaim to zero, and produce a spurious 409 — #1's symptom from the opposite direction. After this change admission does not read that column at all.
5. **Fine-tune jobs join the conflict list for explanation only.** They feed no arithmetic (there is none left to feed). Fresh deploys are already *correct* about training memory — they read the raw node reading, which includes it — they are merely unable to say so.
6. **No `force`/override flag.** The escape hatch already exists and is better: a restart accepts a `gpuMem` config override, so a user who knows the node is tighter than the recipe assumes lowers the request and is admitted on the merits, rather than by disabling the check. On unified memory a bypass produces the OOM-kill path the check exists to prevent.

The vocabulary for all of this now lives in `CONTEXT.md` under **Node memory** (admission, node reading, authorised share, unattributed memory, reclaim). Use those words in code and comments.

### Current behavior

On a restart, the admission orchestrator defines reclaimable memory as *everything in the node reading not attributable to another deployment row*. With no other rows on the node, reclaimable equals the entire node reading, the node's used memory computes to zero, and **the restart is admitted no matter how full the node is**.

Any memory belonging to something the manager has no row for is therefore credited to the restarting deployment as if it were about to be freed. Three ordinary ways to get there, none requiring a mistake:

- a running fine-tune job — training is queried by no admission path and `FineTuneJob` has no memory columns, so a training container holding tens of GiB is invisible;
- a container surviving node offboarding, which deletes the deployment rows but not the containers;
- any hand-started container (a probe, a benchmark).

On GB10 unified memory the admitted deploy then contends and gets OOM-killed (#92).

Fresh deploys are unaffected: they reclaim nothing, so the node reading is used as-is and over-commit is correctly refused. But when unattributed memory causes a *fresh* refusal, the 409 arrives with an empty conflict list — something holds the memory and the product cannot say what (an instance of #101).

### Desired behavior

A restart is credited with at most what it was authorised to hold on that node. Unattributed memory counts against it, so a node holding a training run reads as full rather than empty, and a restart that would not fit is refused.

A refusal names its holders, including fine-tune jobs, so the user knows what to stop.

Accept explicitly: a deployment authorised for more than it takes still over-credits itself (an image recipe declaring no `gpu_memory_utilization` falls back to `0.85` for a workload taking a third of that). This is bounded and honest — a deployment can never credit itself more than a fresh deploy of the same recipe would be allowed to request — where today's behaviour is unbounded. Closing the authorised-vs-actual gap needs measurement that decision 1 rejects; do not attempt it here.

### Key interfaces

- **The pure shortfall function** already clamps its reclaim input and needs no change. All of the fix lands in the Prisma-coupled orchestrator that populates the snapshot.
- **`NodeSnapshot.reclaimableMB`** keeps its name and its clamping, but its *meaning* changes from "everything unexplained" to "at most the authorised share". Its doc comment currently states the old meaning and must be rewritten — it is the thing that will mislead the next reader, exactly as the `#1` comment misled this one.
- **The admission orchestrator's signature** needs the restarting deployment's saved share alongside the requested one. Shape is yours; the constraint is that the two values arrive separately and cannot be confused, and that a caller who passes only one gets the fresh-deploy behaviour.
- **The conflict type** gains fine-tune entries. Both memory fields are `null` for them — nothing consumes those numbers after decision 4. Nothing downstream parses the conflict list structurally (no dashboard reader, no OpenAPI schema); the rendered sentence is the whole surface, so adding entries breaks no contract.
- **Fine-tune holding statuses.** There is no existing constant; `ACTIVE_DEPLOYMENT_STATUSES` is the only precedent. A job holds memory across **three** columns, not one:
  - `status` — `pending`, `starting`, `running`, `stopping` (terminal: `completed`, `failed`). Include `pending` to match the deployment precedent, which treats a job about to hold memory as holding it.
  - `mergeStatus` — `running`. Merging loads the base model and is easily the largest allocation a job makes.
  - `quantizationStatus` — `quantizing`.
  Multi-node training runs have per-node rows analogous to a deployment's cluster rows; a worker node is held by the job just as the head is.
- **The deploy path** persists the resolved `gpu_memory_utilization` into the deployment's config blob so it stops being implicit. Every current deployment runs at the recipe default with nothing stored, so also keep the existing fallback chain (live recipe default, then `0.85`) for rows predating this change.
- **The refusal message** should make the `gpuMem` lever discoverable, since decision 6 makes it the only way through a refusal a user disagrees with.

### Acceptance criteria

- [ ] **Property:** for any snapshot, the used memory after reclaim is never lower than the node reading minus the restarting deployment's authorised share. Stated as a plain-English invariant above the property, per the repo's testing conventions.
- [ ] **Integration (the reported bug):** a node whose latest metric shows memory held by a running `FineTuneJob`, with no active deployment rows besides the one restarting, refuses a restart that does not fit. This test fails on `main`.
- [ ] **Integration (#1 does not regress):** a resident cluster deploy restarting at a *higher* `gpu_memory_utilization` is still admitted. The existing test for this must pass unmodified.
- [ ] **Integration (the latent 409):** two active deployments on one node, each carrying a node-wide figure in its memory column, do not spuriously refuse a restart of either.
- [ ] **Integration:** a restart supplying a higher `gpuMem` override is credited only for its *saved* share — the override sizes the request, never the reclaim.
- [ ] **Integration:** a fresh deploy's behaviour is unchanged — over-commit still refused, node reading used as-is.
- [ ] A new deployment's persisted config carries its resolved `gpu_memory_utilization`; a deployment without one still admits correctly via the fallback chain.
- [ ] A fine-tune job holding a node appears by name and status in the rendered refusal message, distinguishable from a deployment, for all three holding columns.
- [ ] The stale doc comment describing the old reclaim rule is replaced, and the `#1` reference is kept with a note on what it got right and what it missed — the reasoning was sound, the unconsidered case was unattributed memory.
- [ ] `npm test` green.

### Out of scope

- **Any change under `packages/agent/`.** This is server-only. No per-process or per-container memory attribution, no agent version bump.
- **Making the deployment memory column honest for vLLM.** It stays a node sample; after this change admission no longer depends on it. Renaming it is a separate cleanup.
- **A `force` / override flag** (decision 6).
- **VRAM admission for *starting* a fine-tune job.** Training runs today with no admission check at all — a real gap, and a separate issue. This change only makes training *visible* to deployment admission.
- **Giving fine-tune jobs a memory figure**, on the row or in a recipe. Decision 5.
- **#101's disk-breakdown work.** This closes one instance of the same complaint; it does not generalise it.

**Comment by kreuzhofer** [comment]
> *This was generated by AI during triage.*

The stance behind the brief's decisions 1–6 is now written down as **ADR 0004 — The manager measures nodes, never deployments** (`docs/adr/0004-node-memory-attribution.md`), with the vocabulary in `CONTEXT.md` § Node memory.

Read the ADR before the brief if anything in the decisions list looks backwards — particularly "bound by the authorised share, not by the recorded actual", which reads wrong until you know that the recorded actual is the whole node's reading copied onto the deployment row.

The ADR also records two things the brief only implies:

- **why** per-process attribution is rejected rather than merely deferred (a CUDA-context census cannot see the page-cache pressure that killed the replica in #92, so it is incomplete by construction, not by effort);
- **what stays broken on purpose** — training runs are still admitted with no VRAM check of their own. This work makes training *visible* to deployment admission; it does not give training admission.

**Comment by kreuzhofer** [comment]
Fixed in f36438c.

## What landed

A restart is now credited at most its **authorised share** of a node (`round(node.vramTotal × its saved gpu_memory_utilization)`), so unattributed memory — a training run, a container surviving an offboarded node, a hand-started benchmark — counts *against* it. `reclaimForRestart` in `packages/server/src/admission/vram.ts` is the pure seam; `computeVramShortfall` needed no change.

- The `otherFootprintMB` subtraction is **deleted**, per ADR 0004 Decision 3. Admission no longer reads `Deployment.vramActual` / `vramEstimate` at all.
- The orchestrator takes the two shares separately and unconfusably: `checkVllmVramAdmission(nodeIds, requestedGpuMemUtil, { deploymentId, authorisedGpuMemUtil })`. Omit the third argument and you get fresh-deploy behaviour.
- Both deploy paths persist the resolved share as `config.authorisedGpuMem`, written *after* `...config` and added to the restart route's `RESERVED` set — a caller able to set it would be authoring its own reclaim allowance.
- Fine-tune jobs now appear among the holders a refusal names, across all three columns they hold memory in (`status`, `mergeStatus: running` → "merging", `quantizationStatus: quantizing`), on worker nodes as well as heads, via one `fineTuneHoldingStatus` rule. They feed no arithmetic.
- The refusal names the `gpuMem` lever, since Decision 5 makes it the only way through a refusal a user disagrees with.
- #1's comment is kept, with what it got right and what it missed.

## Acceptance criteria

All ten met. Notable details:

- The **#1 regression test passes unmodified**, as required.
- The neighbouring test *"still 409s when the resident VRAM is held by a DIFFERENT deployment"* **was rewritten**: its mechanism (the other-deployment subtraction) is the thing Decision 3 deletes. It now asserts the same intent through the new bound — the restarting deployment is given a real 0.05 share, is credited 6 GB rather than the node's 95 GB, and is still refused with `big-resident` named.
- Two review findings were fixed before commit. A **no-op restart downgraded a persisted 0.88 share to 0.85**, because the restart's *request* chain reads only the sparkrun catalog and falls back to 0.85 for dgxrun and fine-tune rows — re-arming #1's under-credit. The write-back is now gated on an explicit `overrides.gpuMem`, and `savedConfig.dgxrunRecipe.defaults.gpu_memory_utilization` joins the *authorised* chain so pre-existing dgxrun rows recover their real share. Regression-tested in `deployments.dgxrun.test.ts` (verified failing without the fix: `expected 0.85 to be 0.88`).

## Verification

`npm test` green — 158 files, 1563 tests. Server-only; nothing under `packages/agent/` was touched, so no agent version bump.

**Not verified on hardware.** No restart was exercised against a real Spark holding a live training run. The behaviour is covered by integration tests against a per-suite SQLite, which is the level this admission path is testable at; the first real refusal on a node that is also training will look like a regression and is not one (ADR 0004 § Consequences).

## Follow-ups filed separately

The two gaps I deliberately left rather than widening this change — see the linked issues below.

**Comment by kreuzhofer** [comment]
Follow-ups: #122 (dgxrun restarts admit against 0.85 while launching at the recipe's share) and #123 (fine-tune rows predating this recover a flat 0.85). Both are fail-closed or refusing-direction, which is why neither was folded in here.
